File: examples/v0.1.0/demo-agents/python/demo_agent.py
#!/usr/bin/env python3
"""
Demo agent: generic consumer of AOM example outputs.
Input: --folder, --test-case (full stem from list_test_cases(folder)).
Output: AOM output JSON (examples/<folder>/outputs/*.output.json).
"""
import json
import sys
from pathlib import Path

_SCRIPT_DIR = Path(__file__).resolve().parent
# Repo root is four levels up from this file: examples/v0.1.0/demo-agents/python
REPO_ROOT = _SCRIPT_DIR.parent.parent.parent.parent


def _examples_dir():
    for name in ("Examples", "examples"):
        d = REPO_ROOT / name
        if d.is_dir():
            return d
    return REPO_ROOT / "examples"


def _section_dir(folder: str) -> Path:
    return _examples_dir() / folder.strip("/").replace("\\", "/").rstrip("/")


def _outputs_dir(folder: str) -> Path:
    return _section_dir(folder) / "outputs"


def _load_json(path: Path) -> dict:
    return json.loads(path.read_text(encoding="utf-8-sig"))


def _create_outputs_hint(folder: str) -> str:
    # Hint keeps the command simple; CLI wrapper aom.py uses the same script.
    return "Run: python aom.py create-outputs"


def list_folders() -> list[str]:
    """Folder paths (relative to examples/) that have outputs/ with at least one *.output.json."""
    ex = _examples_dir()
    if not ex.is_dir():
        return []
    folders = []
    # Walk recursively under examples/, but record the parent of each outputs/ directory
    # as the logical folder id (e.g. "v0.1.0/login-single").
    for out_dir in ex.rglob("outputs"):
        if not out_dir.is_dir():
            continue
        # Skip demo-agents or hidden paths
        if any(part.startswith(".") for part in out_dir.parts):
            continue
        if "demo-agents" in out_dir.parts:
            continue
        if not any(out_dir.glob("*.output.json")):
            continue
        folder_rel = out_dir.parent.relative_to(ex).as_posix()
        if folder_rel and folder_rel not in folders:
            folders.append(folder_rel)
    return sorted(folders)


def list_test_cases(folder: str) -> list[str]:
    """All output test case ids (file stems) for the folder. Includes outputs that exist and those marked only with .skip (hand-edited)."""
    out_dir = _outputs_dir(folder)
    if not out_dir.is_dir():
        return []
    cases = {f.stem for f in out_dir.glob("*.output.json")}
    # Include stems from .skip markers so hand-edited outputs are still tested (same count as create-outputs)
    for f in out_dir.glob("*.output.json.skip"):
        stem = (f.name[:-5] if f.name.endswith(".skip") else f.name)  # remove ".skip"
        cases.add(Path(stem).stem)  # e.g. "x.output.json" -> "x.output"
    return sorted(cases)


def _resolve_output_path(outputs_dir: Path, test_case: str) -> Path | None:
    test_case = test_case.strip()
    p = outputs_dir / f"{test_case}.json"
    if p.is_file():
        return p
    p = outputs_dir / f"{test_case}.output.json"
    if p.is_file():
        return p
    # Fallback for hand-edited outputs that only exist as .output.json.skip markers
    p = outputs_dir / f"{test_case}.json.skip"
    if p.is_file():
        return p
    return None


def resolve_test_case(folder: str, test_case: str) -> str | None:
    """Resolve folder + test_case (exact or prefix) to full stem, or None if invalid. Use to validate before adding to test set (0 tests when invalid)."""
    if not folder or not test_case:
        return None
    folder = folder.strip("/").replace("\\", "/").rstrip("/")
    folders = list_folders()
    if folder not in folders:
        return None
    valid = list_test_cases(folder)
    if not valid:
        return None
    t = test_case.strip()
    if t in valid:
        return t
    match = [c for c in valid if c.startswith(t + ".")]
    return match[0] if len(match) == 1 else None


def run(*, folder: str, test_case: str) -> dict:
    """Return the pre-generated output for the given folder and test case (full stem)."""
    folder = folder.strip("/").replace("\\", "/").rstrip("/")
    if not folder:
        raise ValueError("folder is required")
    outputs_dir = _outputs_dir(folder)
    if not outputs_dir.is_dir():
        raise FileNotFoundError(
            f"Outputs dir not found: {outputs_dir}. {_create_outputs_hint(folder)}"
        )
    if not list_test_cases(folder):
        raise FileNotFoundError(
            f"No output files in {outputs_dir}. {_create_outputs_hint(folder)}"
        )
    resolved = resolve_test_case(folder, test_case)
    path = _resolve_output_path(outputs_dir, resolved) if resolved else None
    if path is None:
        raise ValueError(f"Unknown test_case or missing output file: {test_case!r}")
    return _load_json(path)


def main():
    import argparse
    p = argparse.ArgumentParser(description="Demo agent: folder + test-case → output")
    p.add_argument("--folder", metavar="NAME", required=True, help="Example folder under examples/ (e.g. v0.1.0/login-single, v0.1.0/ecom-flow)")
    p.add_argument("--test-case", metavar="NAME", required=True, help="Test case id (full stem from list_test_cases)")
    args = p.parse_args()
    try:
        out = run(folder=args.folder, test_case=args.test_case)
        print(json.dumps(out, indent=2))
    except (FileNotFoundError, ValueError) as e:
        print(str(e), file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: examples/v0.1.0/demo-agents/node/demo_agent.js
#!/usr/bin/env node
/**
 * Demo agent: generic consumer of AOM example outputs.
 * Input: --folder, --test-case (full stem from listTestCases(folder)).
 * Output: AOM output JSON (examples/<folder>/outputs/*.output.json).
 */

import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
// Repo root is four levels up: examples/v0.1.0/demo-agents/node -> ... -> repo root
const REPO_ROOT = path.resolve(__dirname, "..", "..", "..", "..");

function examplesDir() {
  for (const name of ["Examples", "examples"]) {
    const d = path.join(REPO_ROOT, name);
    if (fs.existsSync(d) && fs.statSync(d).isDirectory()) return d;
  }
  return path.join(REPO_ROOT, "examples");
}

function outputsDir(folder) {
  const norm = folder.replace(/^[/\\]+|[/\\]+$/g, "");
  return path.join(examplesDir(), norm, "outputs");
}

function loadJson(filePath) {
  return JSON.parse(fs.readFileSync(filePath, "utf-8"));
}

export function listFolders() {
  const ex = examplesDir();
  if (!fs.existsSync(ex) || !fs.statSync(ex).isDirectory()) return [];
  const folders = new Set();

  function walk(dir) {
    const entries = fs.readdirSync(dir, { withFileTypes: true });

    // If this dir has an outputs/ child with *.output.json, record it
    const outputsChild = entries.find((e) => e.isDirectory() && e.name === "outputs");
    if (outputsChild) {
      const outDir = path.join(dir, outputsChild.name);
      try {
        if (fs.readdirSync(outDir).some((f) => f.endsWith(".output.json"))) {
          const rel = path.relative(ex, dir).replace(/\\/g, "/");
          if (rel && !rel.split("/").some((p) => p.startsWith(".") || p === "demo-agents")) {
            folders.add(rel);
          }
        }
      } catch (_) {}
    }

    // Recurse into child directories (excluding outputs/ itself and demo-agents/ or hidden folders)
    for (const entry of entries) {
      if (!entry.isDirectory()) continue;
      if (entry.name === "outputs" || entry.name === "demo-agents" || entry.name.startsWith(".")) continue;
      walk(path.join(dir, entry.name));
    }
  }

  walk(ex);
  return [...folders].sort();
}

export function listTestCases(folder) {
  const outDir = outputsDir(folder);
  if (!fs.existsSync(outDir) || !fs.statSync(outDir).isDirectory()) return [];
  const cases = new Set();
  for (const name of fs.readdirSync(outDir)) {
    if (name.endsWith(".output.json")) cases.add(path.basename(name, ".json"));
    if (name.endsWith(".output.json.skip")) cases.add(path.basename(path.basename(name, ".skip"), ".json")); // stem e.g. _login.escalated.output
  }
  return [...cases].sort();
}

function resolveOutputPath(outDir, testCase) {
  const t = String(testCase).trim();
  const p1 = path.join(outDir, t + ".json");
  if (fs.existsSync(p1) && fs.statSync(p1).isFile()) return p1;
  const p2 = path.join(outDir, t + ".output.json");
  if (fs.existsSync(p2) && fs.statSync(p2).isFile()) return p2;
  // Fallback for hand-edited outputs that only exist as .output.json.skip markers
  const p3 = path.join(outDir, t + ".json.skip");
  if (fs.existsSync(p3) && fs.statSync(p3).isFile()) return p3;
  return null;
}

/**
 * Resolve a folder + test_case (exact or prefix) to a full stem, or null if invalid.
 * Use to validate before adding to test set (0 tests when invalid).
 */
export function resolveTestCase(folder, testCase) {
  if (!folder || !testCase) return null;
  const folderNorm = folder.replace(/^[/\\]+|[/\\]+$/g, "");
  const folders = listFolders();
  if (!folders.includes(folderNorm)) return null;
  const cases = listTestCases(folderNorm);
  if (cases.length === 0) return null;
  const t = String(testCase).trim();
  if (cases.includes(t)) return t;
  const match = cases.filter((c) => c.startsWith(t + "."));
  return match.length === 1 ? match[0] : null;
}

/**
 * @param {{ folder: string, test_case: string }} opts
 * @returns {object} AOM output
 */
export function run(opts = {}) {
  const { folder, test_case: testCase } = opts;
  if (!folder || !testCase) throw new Error("folder and test_case are required");
  const folderNorm = folder.replace(/^[/\\]+|[/\\]+$/g, "");
  const outDir = outputsDir(folderNorm);
  if (!fs.existsSync(outDir) || !fs.statSync(outDir).isDirectory()) {
    throw new Error(`Outputs dir not found: ${outDir}. Run: node aom.mjs create-outputs`);
  }
  const cases = listTestCases(folderNorm);
  if (cases.length === 0) {
    throw new Error(`No output files in ${outDir}. Run: node aom.mjs create-outputs`);
  }
  const resolved = resolveTestCase(folderNorm, testCase);
  const filePath = resolved ? resolveOutputPath(outDir, resolved) : null;
  if (!filePath) throw new Error(`Unknown test_case or missing output file: ${testCase}`);
  return loadJson(filePath);
}

function main() {
  const args = process.argv.slice(2);
  let folder = null;
  let testCase = null;
  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--folder" && args[i + 1]) {
      folder = args[++i];
    } else if (args[i] === "--test-case" && args[i + 1]) {
      testCase = args[++i];
    }
  }
  try {
    if (!folder || !testCase) {
      console.error("Usage: node demo_agent.js --folder NAME --test-case STEM");
      process.exit(1);
    }
    const out = run({ folder, test_case: testCase });
    console.log(JSON.stringify(out, null, 2));
  } catch (e) {
    console.error(e.message);
    process.exit(1);
  }
}

if (process.argv[1]?.endsWith("demo_agent.js")) {
  main();
}
